doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,250 @@
+   As a result, in addition to `Strict`, one may need to also use one or more of the following GHC flags: `-fno-strictness`, `-fno-specialise`, `-fno-spec-constr`, `-fno-unbox-strict-fields`, `-fno-unbox-small-strict-fields`.
+   Not only is it inconvenient, but failing to do so may lead to obscure compilation errors.
+2. Enable `Strict` automatically.
+   This can be done in a GHC driver plugin.
+   But like said in Option 1, one doesn't usually want to enable `Strict` on off-chain code.
+   Thus at the very least, we'd need to provide the option to turn it off via `NoStrict`, and this is tricky to implement.
+3. Introduce the following function:
+
+  ```haskell
+  nonstrict :: forall a. a -> a
+  nonstrict = id
+  {-# NOINLINE nonstrict #-}
+  ```
+
+  Then, to create a non-strict let-binding, we write
+
+  ```haskell
+  let x = nonstrict rhs in body
+  ```
+
+  Because of the NOINLINE pragma, `nonstrict` is guaranteed to be visible to the Plutus Tx compiler.
+  Whenever the Plutus Tx compiler sees `let x = nonstrict rhs`, it first compiles `rhs` normally; then it creates a nonstrict binding in PIR: `let ~x = rhs_compiled`.
+
+The advantages of making let-bindings strict by default, and using `nonstrict` for non-strict bindings are:
+
+- It makes it more convenient, rather than less convenient, to create strict bindings than non-strict bindings, which is what a strict language should do.
+- It works for both on-chain code and off-chain code. A let-binding is strict by default when used in on-chain code, and non-strict by default otherwise.
+
+The disadvantages are:
+
+- It is a magic function that needs special support in the Plutus Tx compiler.
+- This is a critical one: it is susceptible to GHC inlining bindings unconditionally.
+  If GHC decides to turn `let x = rhs in ...x...` into `...rhs...`, then `x` effectively becomes non-strict, and there's nothing the Plutus Tx compiler can do to recover `x`.
+
+If we can modify the Plutus Tx compiler such that we can turn off GHC's pre-inliner (which will be discussed later), then option 3 becomes the most preferrable.
+Otherwise option 3 doesn't really work and we'll have to go with option 1.
+
+### Defining non-strict functions
+
+In Plutus Tx, like most other strict languages, only a few built-in constructs are non-strict (such as if-then-else).
+It is often cumbersome, if not impossible, to create user-defined non-strict functions.
+
+It would be useful to allow user-defined non-strict functions, so that we can define functions like `&&` and `||` that short-circuit properly.
+Some languages like Scala have special syntaxes for defining non-strict functions; we can't change Haskell's syntax, so we need to find a different solution.
+
+The idea is to make the Plutus Tx compiler perform inlining and beta reduction for functions with INLINE pragmas, i.e., the same as how GHC treats INLINE pragmas.
```

Comment:
I'm sorry, I still hate it :grimacing: I _really_ don't want to use `INLINE` for this. I'd be more happy to use a custom source annotation as previously discussed, although I still very much don't love it.